### ISSUE TITLE
Kira-D's GSOC Work

### ISSUE THREAD

**Issue by sushain97** [goavki/apertium-html-tools#58]
Hi @Kira-D, here are most of the improvements we currently have on our wishlist for Apertium's website, some editorial comments on them and links to relevant places. I've tried to sort it in roughly increasing difficulty and categorized it a bit. I would recommend starting with some of the easier ones to gain a familiarity with the codebase. I've bolded three of the most significant improvements. Of them, website translation is the most straightforward and already has an almost complete interface. Translation suggestions are the next priority. Picking from this list to create a specific work plan might be a good idea :)
### Enhancements
- [ ] Installation notification when we have several timeouts in a row (#9)
  - This one has a very high benefit to cost ratio, I would really like to see it implemented.
  - The number of timeouts needed to trigger this notification is ideally a setting. However, if this is too complicated to achieve given the fairly small benefit, that's fine.
  - Design/placement/UI is pretty important here.
- [ ] CDN script hashes (#50)
  - This really isn't a priority at all, mostly something to tackle if you're interested.
  - Look into compatibility, as Tino mentions.
- <s> Localisation language dropdown sorting (#23)
  - This one requires a bit of research and discussion.
  - Looking into how different sites (e.g. Google) handle this may be instructive.</s>
- [ ] **Website translation**
  - This feature has been often requested and used to be present on the old website. As such, it's a great starting point for the larger improvements.
  - There's already been a fair bit of work done towards this (such as #41). Unfortunately, there's been a bit too much ([websiteTranslation](https://github.com/goavki/apertium-html-tools/tree/websiteTranslation) branch). Look at both interfaces and take the best one/combine them. I'm guessing that the one from #41 will be better but I could easily be wrong.
  - Make sure that the head of APY's translation endpoint still works with this (most likely will).
  - Address both Unhammer's suggestions from #41. There may be need for some minor UI adjustments when all is said and done.
  - Make sure to send unique events to Piwik for this (this may already be implemented)
  - Feel free to create a new PR/issue for this if you want a place to discuss.
- [ ] **Translation suggestions** (#55)
  - This functionality is highly desired, very useful and fairly commonplace.
  - Significant interface work has already been done. Start from the [suggestions](https://github.com/goavki/apertium-html-tools/tree/suggestions) branch.
  - With regards to UI refinements, play with the UI (this also applies to more testing needed)! Improve what you think can be improved and then we'll do the same (i.e. play with it and offer suggestions). Let us know when _you_ think it's good to go.
  - Giving the ReCAPTCHA instance the current localisation would be awesome. However, I'm not sure if we even need to do this or whether Google will automagically handle it (test it!)
  - In terms of tests, start with tests that ensure that a) suggestions are being recorded and b) suggestions are only recorded with valid ReCAPTCHA tokens. 
- [ ] **Spell checking** (#12)
  - Interface is probably the most tricky part here.
  - Requires modifications to APY (should be straightforward since Apertium modes exist for it).
  - There's an interface in the [spellChecker](https://github.com/goavki/apertium-html-tools/tree/spellChecker) branch that can be improved on and is already pretty functional. Don't start from scratch unless it's really worth it!
- [ ] **Translation Confidence Estimation**
- [ ] **Dictionary Mode**
  - A nice front end needs to be developed for this, with nice probability bars if there is probability data available from the back end
  - A back end also needs to be developed
    - this back end needs to be able to make use of some format of data (probably a database table), which could include word pairs, and optionally also probabilities.
- [ ] **Language Detection**
  - There will need to be a specific algorithm or pipeline developed for the back end of this, to be researched and discussed with us.
  - Most of the front-end and server-to-interface pipeline is in place, but there should be some sort of "did you mean ___?" message, like for other sites that have language detection.
### Bugs
- [ ] Subtitle display in RTL localisations (#18)
  - This is probably pretty simple to address.
  - @jonorthwash would probably really like it if this were fixed.
- [ ] Dropdown behaving funky in RTL localisations (#19)
  - Annoying bug that I would look into when I get frustrated with another task :P
- [ ] Character set messing up (#48)
  - Work with @jonorthwash to ensure this issue is solved.
  - Make sure to test in the same environment/context he's in.
- [ ] "Translation not yet available" when simply no target language selected (#38)
  - This may or may not still be an issue.
### Somewhere in Between
- [ ] Font issues (#17)
  - The best solution may be switching fonts based on what language is being used. This could start getting tricky (might require storing more information in the localization JSON)
  - Google Fonts has fonts that support many different scripts, this is worth looking into. There might be one font that we can switch between loading different versions of.
- [ ] Documentation improvements (#20 partially)
  - The documentation for html-tools (and APY) has gotten slowly out of date. It could use gradual fixing. Another task probably best done in little blocks.
- [ ] Input box and output area positioning in RTL languages (#49)
  - Discuss with @jonorthwash exactly what the behavior we want here is.
### General notes
- Work in branches
- Pay attention to code style (Sushain, me, can be a stickler about it) 
- Pay attention to code style (heh.)

**Comment by sushain97** [comment]
Closing this since it's outdated and this GSOC project is over. Issues for each one exist that can be referenced.
